tests/test_pipeline.py

```diff
@@ -275,6 +316,90 @@ def test_gridsearchcv(self) -> None:
             for k, value in param_grid.items():
                 self.assertIn(grid_search_cv.best_params_[k], value)
 
+    def test_caching(self) -> None:
+        """Test if the caching gives the same results and is faster on the second run."""
+
+        molecule_net_logd_df = pd.read_csv(
+            "https://deepchemdata.s3-us-west-1.amazonaws.com/datasets/Lipophilicity.csv"
+        ).head(1000)
+        pipeline = _get_rf_regressor()
+        mem = Memory(location="./cache", verbose=0)
+        pipeline.memory = mem
+        # Run fitting 1
+        start_time = time.time()
+        pipeline.fit(
+            molecule_net_logd_df["smiles"].tolist(),
+            molecule_net_logd_df["exp"].tolist(),
+        )
+        time1 = time.time() - start_time
+        # Get predictions
+        pred1 = pipeline.predict(molecule_net_logd_df["smiles"].tolist())
+
+        # Reset the last step with an untrained model
+        pipeline.steps[-1] = ("rf", RandomForestRegressor(random_state=42, n_jobs=-1))
+
+        # Run fitting 2
+        start_time = time.time()
+        pipeline.fit(
+            molecule_net_logd_df["smiles"].tolist(),
+            molecule_net_logd_df["exp"].tolist(),
+        )
+        time2 = time.time() - start_time
+        # Get predictions
+        pred2 = pipeline.predict(molecule_net_logd_df["smiles"].tolist())
+
+        # Compare results
+        self.assertTrue(np.allclose(pred1, pred2))
+        self.assertLess(time2, time1)
```

Comment:
Testing timings in unit tests is considered a bad thing. Timings depend on many external circumstances. Our code can be correct even when these external circumstances lead to bad runtimes. Instead, I would just test to ensure that the results are correct. 

You could also add a test that tests that the cached element is used and not recreated every time the pipeline is called. Maybe wrap the featurization-element in a Wrapper class that counts executions of the fit_transform method, like:

```Python

class MyWrapper:

def __init__(element):
  self.counter = 0
  self.element = element

def fit_transform(X, y):
  self.counter += 1
  return self.element.fit_transform(X, y)

featurization_element_wrapped = MyWrapper(MorganFP())
pipeline = Pipeline([
  ..,
  ("featurizater", featurization_element_wrapped),
 ..]
pipeline.fit(X,y)

self.assertEqual(featurization_element_wrapped.counter, 1)
```

Note that this should only work when you don't parallelize the execution.

Reply:
Implemented. I had to make the `counter` var a global var, as otherwise the hash of the object would be changed. It is kind of ugly, do you have a better idea?